users/src/test/java/io/crate/metadata/UsersPrivilegesMetaDataTest.java

```diff
@@ -72,16 +94,94 @@ public void testXContent() throws IOException {
         // reflects the logic used to process custom metadata in the cluster state
         builder.startObject();
 
-        UsersPrivilegesMetaData usersPrivilegesMetaData = new UsersPrivilegesMetaData(usersPrivileges);
         usersPrivilegesMetaData.toXContent(builder, ToXContent.EMPTY_PARAMS);
         builder.endObject();
 
         XContentParser parser = JsonXContent.jsonXContent.createParser(builder.bytes());
         parser.nextToken(); // start object
-        UsersPrivilegesMetaData usersPrivilegesMetaData2 = (UsersPrivilegesMetaData)new UsersPrivilegesMetaData().fromXContent(parser);
+        UsersPrivilegesMetaData usersPrivilegesMetaData2 = (UsersPrivilegesMetaData) new UsersPrivilegesMetaData().fromXContent(parser);
         assertEquals(usersPrivilegesMetaData, usersPrivilegesMetaData2);
 
         // a metadata custom must consume its surrounded END_OBJECT token, no token must be left
         assertThat(parser.nextToken(), nullValue());
     }
+
+    @Test
+    public void testApplyPrivilegesSameExists() throws Exception {
+        long rowCount = usersPrivilegesMetaData.applyPrivileges(USERNAMES, new HashSet<>(PRIVILEGES));
+        assertThat(rowCount, is(0L));
+    }
+
+    @Test
+    public void testRevokeWithoutGrant() throws Exception {
+        long rowCount = usersPrivilegesMetaData.applyPrivileges(
+            Collections.singletonList("Arthur"),
+            Collections.singletonList(REVOKE_DCL)
+        );
+        assertThat(rowCount, is(0L));
+        assertThat(usersPrivilegesMetaData.getUserPrivileges("Arthur"), is(PRIVILEGES));
+    }
+
+    @Test
+    public void testRevokeWithGrant() throws Exception {
+        long rowCount = usersPrivilegesMetaData.applyPrivileges(
+            Collections.singletonList("Arthur"),
+            Collections.singletonList(REVOKE_DML)
+        );
+
+        assertThat(rowCount, is(1L));
+        assertThat(usersPrivilegesMetaData.getUserPrivileges("Arthur"), contains(GRANT_DQL));
+    }
+
+
+    @Test
+    public void testRevokeWithGrantOfDifferentGrantor() throws Exception {
+        long rowCount = usersPrivilegesMetaData.applyPrivileges(
+            Collections.singletonList("Arthur"),
+            Collections.singletonList(new Privilege(Privilege.State.REVOKE, Privilege.Type.DML, Privilege.Clazz.CLUSTER, null, "hoschi"))
+        );
+
+        assertThat(rowCount, is(1L));
+        assertThat(usersPrivilegesMetaData.getUserPrivileges("Arthur"), contains(GRANT_DQL));
+    }
+
+    @Test
+    public void testDenyGrantedPrivilegeForUsers() throws Exception {
+        long rowCount = usersPrivilegesMetaData.applyPrivileges(
+            USERNAMES,
+            Collections.singletonList(DENY_DQL)
+        );
+        assertThat(rowCount, is(2L));
```

Comment:
shouldn't the deny also be stored and the grant removed? if so we should assert that imho